man/kmrgenscript.1
'\"
.\" Copyright (C) 2012-2018 RIKEN R-CCS
.\" nroff -u0 -Tlp -man
.TH kmrgenscript.py 1 "2014-02-04" "KMR" "User Commands"
.SH NAME
kmrgenscript.py \- generates a job-script
.SH SYNOPSIS
.LP
.nf
\fBkmrgenscript.py\fR [-e node] [-p prefix] [-o outfile] [-d dir] [-O outdir] [-t rsctime] [-S sched] [-w file] [-M] [-f] -m mapper -r reducer
.fi
.SH DESCRIPTION
.sp
.LP
kmrgenscript.py generates a job-script for job schedulers.  It reads a
template of a job-script "kmrgenscript.template" either from the
current directory, "lib" directory next to "bin" where this script is
found, or installation directory, searching in this order.  Currently,
it only supports "Parallelnavi" on K as a job scheduler.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
-e node
.ad
.sp .6
.RS 4n
Specifies the number of nodes (processes) to execute.  Default is 1.
.RE
.sp
.ne 2
.na
-p prefix
.ad
.sp .6
.RS 4n
Specifies a prefix to the input file names.  Default is "part".
.RE
.sp
.ne 2
.na
-o outfile
.ad
.sp .6
.RS 4n
Specifies a prefix to the output file names.  Default is "output".
.RE
.sp
.ne 2
.na
-d indir
.ad
.sp .6
.RS 4n
Specifies the input directory.  Default is the current directory.
.RE
.sp
.ne 2
.na
-O outdir
.ad
.sp .6
.RS 4n
Specifies the output directory.  The output directory holds the result
files.  Default is the current directory.
.RE
.sp
.ne 2
.na
-t rsctime
.ad
.sp .6
.RS 4n
Specifies time limit in the job resources.  This is given by
"00:00:00" format.
.RE
.sp
.ne 2
.na
-S sched
.ad
.sp .6
.RS 4n
Specifies a job scheduler.  Only "K" for Parallelnavi on K is
supported.
.RE
.sp
.ne 2
.na
-w scrfile
.ad
.sp .6
.RS 4n
Specifies a script output file.  Default is STDOUT.
.RE
.sp
.ne 2
.na
-M
.ad
.sp .6
.RS 4n
Specifies multiple input files are given to a mapper.  In case of the
number of input files is greater than the number of processes, it
assigns multiple files to one process.  For using this option, -d
option is required.
.RE
.sp
.ne 2
.na
-f
.ad
.sp .6
.RS 4n
Forces to create an output directory if it does not exist.
.RE
.SH EXAMPLES
.sp
.LP
For example, the following script is generated.
.sp
.nf
.RS
#!/bin/bash -x
#
#PJM --rsc-list "node=2"
#PJM --rsc-list "elapse=00:10:00"
#PJM --rsc-list "proc-core=unlimited"
#PJM --stg-transfiles "all"
#PJM --mpi "use-rankdir"
#PJM --stgin "rank=* ./kmrshell     %r:./"
#PJM --stgin "rank=* ./mapper       %r:./"
#PJM --stgin "rank=* ./kmrshuffler  %r:./"
#PJM --stgin "rank=* ./reducer      %r:./"
#PJM --stgin "rank=* ./part%r       %r:./input"
#PJM --stgout "rank=* %r:./output.%r ./output.%r"
#PJM -S
\&. /work/system/Env_base
mpiexec -n 2 -of-proc output ./kmrshell ./mapper ./reducer ./input
.fi
.RE


CHANGELOG.md
# CHANGELOG

The list is sketchy.

## 1.9 (2018-08-27)

- 1.9 is a minor update
- Move GitHub repository from pf-aics-riken to riken-rccs
- Change the copyright holder name: RIKEN AICS to RIKEN R-CCS

## 1.8.1 (2016-04-25)

- Add ex/kmreprun.py, an example of simple shell command invoker
- Improve support of Mac OS X
- Several small bug fixes

## 1.8.0 (2015-12-25)

- Add a function tracing option for KMRViz, a KMR visualizer
- Add a simple scan function, kmr_scan_on_value()
- Add a load-leveling shuffle, kmr_shuffle_leveling_pair_count()
- Add a new kmr example, pagerank.c
- BUGFIX: kmr_match(), kmr_histogram_count_by_ranks(), kmr_find_key()

## 1.7.1 (2015-06-29)

- BUGFIX: kmr_take_one()
- BUGFIX: kmrrungenscript.py

## 1.7 (2015-06-22)

- Change license to BSD
- Strictly check options passed to kmr functions
- Add support for FOCUS supercomputer in kmrrungenscript.py
- Add Python API

## 1.6 (2015-04-01)

- Disable keep_open option in kmr functions except mapper functions
- Support no-fsync mode in Checkpoint/Restart
- Improve performance of KMRRUN when many input files are given
- BUGFIX: Checkpoint/Restart
- EXPERIMENTAL: Support selective mode that reduces IO overhead in
  Checkpoint/Restart

## 1.5 (2015-02-04)

- Reduce overhead of KMRRUN
- Temporal files for reducers are generated in subdirectories when
  using KMRRUN
- WORKAROUND: Abort on calling kmr_read_file_xxx with some MPI
  implementations
- WORKAROUND: Abort on calling kmr_map_processes and using KMRRUN
  with some Open MPI versions (1.6.3 - 1.8.1)

## 1.4 (2014-12-25)

- Resolved a scalability issue of KMRRUN and map functions that
  spawns child programs

## 1.3.2 (2014-11-07)

- Improve job script generator for KMRRUN

## 1.3.1 (2014-11-04)

- Add a word counting sample of KMRRUN
- BUGFIX: KMRRUN command line option parsing

## 1.3 (2014-09-17)

- Add a new function for affinity-aware file assignment

## 1.2 (2014-06-27)

- Support Checkpoint/Restart
- Add KMRRUN command that executes simple MapReduce workflow

## 1.1 (2013-09-20)

- 1.1 is a minor update

## 1.0 (2013-04-26)

- First release


cmd/kmrgenscript.in.py
#!/usr/bin/python
# Copyright (C) 2012-2018 RIKEN R-CCS

## \file kmrgenscript.in.py KMR-Shell Job-Script Generator.

import sys
import os
from optparse import OptionParser

kmrhome = '@KMRHOME@'

## Checks file existence.
#  If file does not exist, it prints an error message and exit.
#  @param path  file path for check.

def checkexist (path) :
    if not os.path.exists(path) :
        print 'Error: file or dir "%s" is not exist.' % path
        sys.exit()

## Checks path is a directory.
#  If path does not exist, it prints an error message and exit.  Or,
#  it creates a directory with force option.

def checkdir (path, force) :
    if os.path.exists(path) :
        if not os.path.isdir(path) :
            print 'Error: "%s" is not directory.' % path
            sys.exit()
    else :
        if force :
            try :
                os.mkdir(path)
            except IOError :
                print 'Error: could not create "%s".' % path
                sys.exit()
        else :
            print 'Error: directory "%s" is not exist. create it or use -f option.' % path
            sys.exit()

## Generates job-script for K.
#  @param node     number of node to execute.
#  @param infile   input file pathname.
#  @param outfile  output file prefix.
#  @param indir    directory path name that has input files.
#  @param outdir   directory path name that has result output files.
#  @param rsctime  resource time limit.
#  @param mapper   pathname of mapper program.
#  @param reducer  pathname of reducer program.
#  @param multi    multiple input file for one mapper process.
#  @param scrfile  output script file name.

def k_scheduler(node, infile, outfile, indir, outdir, rsctime, mapper, reducer, multi, scrfile) :

    # Read template file.
    template = ''
    if template == '' :
        try :
            template = open('kmrgenscript.template').read()
        except IOError :
            pass
    if template == '' :
        try :
            dir0 = os.path.dirname(os.path.realpath(__file__))
            dir1 = os.path.realpath(dir0 + '/../lib')
            #template = open(dir1 + '/kmrgenscript.template').read()
        except IOError :
            pass
    if template == '' :
        try :
            dir2 = os.path.realpath(kmrhome + '/lib')
            template = open(dir2 + '/kmrgenscript.template').read()
        except IOError :
            pass
    if template == '' :
       print 'Error: could not open job-script template.'
       sys.exit()

    # Stage in section
    ncol = len(str(node -1))
    stginstr = ""
    # Stage in reducer if specified
    if reducer :
        stginstr += '#PJM --stgin "rank=* %s %%r:./"' \
                    % ('./' + os.path.basename(reducer))
    if multi :
        files = os.listdir(indir)
        rank = 0
        for file in files :
            ipath = os.path.join(indir, file)
            if len(stginstr) :
                stginstr += '\n'
            stginstr += '#PJM --stgin "rank=%s %s %%r:./work/"' % (rank, ipath)
            rank = rank +1
            if rank >= node :
                rank = 0
    else :
        if len(stginstr) :
            stginstr += '\n'
        ipath = os.path.join(indir, infile)
        stginstr += '#PJM --stgin "rank=* %s%%0%sr %%r:./input"' % (ipath, ncol)

    # Stage out section
    opath = os.path.join(outdir, outfile)
    stgoutstr = '#PJM --stgout "rank=* %%r:./output.%%r %s.%%0%sr"' % (opath, ncol)

    # program execute section
    if not multi :
        if not reducer :
            execstr = 'mpiexec -n %s -of-proc output ./kmrshell -m %s ./input'\
                 % (node, './' + os.path.basename(mapper))
        else :
            execstr = 'mpiexec -n %s -of-proc output ./kmrshell -m %s -r %s ./input'\
                 % (node, './' + os.path.basename(mapper), './' + os.path.basename(reducer))
    else :
        if not reducer :
            execstr = 'mpiexec -n %s -of-proc output ./kmrshell -m %s ./work'\
                 % (node, './' + os.path.basename(mapper))
        else :
            execstr = 'mpiexec -n %s -of-proc output ./kmrshell -m %s -r %s ./work'\
                 % (node, './' + os.path.basename(mapper), './' + os.path.basename(reducer))


    # replace template keyword using parameter.
    script = template % {'NODE': node, 'RSCTIME': rsctime, 'MAPPER': mapper, 'DATASTGIN': stginstr, 'DATASTGOUT': stgoutstr, 'EXEC': execstr, 'KMRHOME': kmrhome}

    # output script
    if scrfile is None :
        print script
    else :
        out = open(scrfile, "w")
        print >> out, script
        out.close()

## Selects job-scheduler.
#  @param node     number of node to execute.
#  @param infile   input file pathname.
#  @param outfile  output file prefix.
#  @param indir    directory path name that has input files.
#  @param outdir   directory path name that has result output files.
#  @param rsctime  resource time limit.
#  @param mapper   pathname of mapper program.
#  @param reducer  pathname of reducer program.
#  @param multi    multiple input file for one mapper process.
#  @param sched    scheduler.
#  @param scrfile  output script file name.

def selectscheduler(node, infile, outfile, indir, outdir, rsctime, mapper, reducer, multi, sched, scrfile) :
    if sched == 'K' :
        k_scheduler(node, infile, outfile, indir, outdir, rsctime, mapper, reducer, multi, scrfile)
    # for other schedulers...

## kmrgenscript main routine.
#  It works on Python 2.4 or later.

if __name__ == "__main__" :

    usage = "usage: %prog [options] -m mapper [-r reducer]"
    parser = OptionParser(usage)

    parser.add_option("-e",
                      "--number-of-exec-node",
                      dest="node",
                      type="int",
                      help="number of execute node",
                      metavar="number",
                      default=1)

    parser.add_option("-p",
                      "--input-file-prefix",
                      dest="infile",
                      type="string",
                      help="input filename prefix",
                      metavar="'string'",
                      default='part')

    parser.add_option("-o",
                      "--outputfile",
                      dest="outfile",
                      type="string",
                      help="output filename prefix",
                      metavar="'string'",
                      default='output')

    parser.add_option("-d",
                      "--inputdir",
                      dest="indir",
                      type="string",
                      help="input directory",
                      metavar="'string'",
                      default='./')

    parser.add_option("-O",
                      "--outputdir",
                      dest="outdir",
                      type="string",
                      help="output directory",
                      metavar="'string'",
                      default='./')

    parser.add_option("-t",
                      "--resource-time",
                      dest="rsctime",
                      type="string",
                      help="resource time",
                      metavar="'string'",
                      default='00:10:00')

    parser.add_option("-m",
                      "--mapper",
                      dest="mapper",
                      type="string",
                      help="mapper path",
                      metavar="'string'")

    parser.add_option("-r",
                      "--reducer",
                      dest="reducer",
                      type="string",
                      help="reducer path",
                      metavar="'string'")

    parser.add_option("-S",
                      "--scheduler",
                      dest="sched",
                      type="string",
                      help="scheduler (default is 'K')",
                      metavar="'string'",
                      default='K')

    parser.add_option("-w",
                      "--write-scriptfile",
                      dest="scrfile",
                      type="string",
                      help="script filename",
                      metavar="'string'")

    parser.add_option("-M",
                      "--multi-input",
                      dest="multi",
                      action="store_true",
                      help="multi input files to one node",
                      default=False)

    parser.add_option("-f",
                      "--force",
                      dest="force",
                      action="store_true",
                      help="force option",
                      default=False)

    (options, args) = parser.parse_args()

    # check parameters.

    if len(args) > 1 :
        parser.error("missing parameter")
        sys.exit()

    if not options.mapper :
        print "mapper not specified\n"
        sys.exit()

    checkexist(options.indir)
    checkdir(options.outdir, options.force)

    if options.multi :
        if options.indir == "./" :
            print "-M option needs -d (input directory) option.\n"
            sys.exit()
        files = os.listdir(options.indir)
        if len(files) < options.node :
            print 'Node number is greater than number of files in %s.\n' % options.indir
            sys.exit()

    selectscheduler(options.node, options.infile, options.outfile,
                    options.indir, options.outdir, options.rsctime,
                    options.mapper, options.reducer, options.multi,
                    options.sched, options.scrfile)

# Copyright (C) 2012-2018 RIKEN R-CCS
# This library is distributed WITHOUT ANY WARRANTY.  This library can be
# redistributed and/or modified under the terms of the BSD 2-Clause License.
